libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java

```diff
@@ -0,0 +1,134 @@
+    static final String THREAD_NAME = "th_name";
+    static final String PARENT_SPAN_NAME = "p_span_name";
+
+    private final TracingTelemetry tracingTelemetry;
+    private final TracerContextStorage<String, Span> tracerContextStorage;
+    private final Supplier<Level> levelSupplier;
+    private final SpanFactory spanFactory;
+
+    /**
+     * Creates DefaultTracer instance
+     *
+     * @param tracingTelemetry tracing telemetry instance
+     * @param tracerContextStorage storage used for storing current span context
+     * @param levelSupplier configured level supplier
+     */
+    public DefaultTracer(
+        TracingTelemetry tracingTelemetry,
+        TracerContextStorage<String, Span> tracerContextStorage,
+        Supplier<Level> levelSupplier
+    ) {
+        this.tracingTelemetry = tracingTelemetry;
+        this.tracerContextStorage = tracerContextStorage;
+        this.levelSupplier = levelSupplier;
+        this.spanFactory = new SpanFactory(levelSupplier, tracingTelemetry);
+    }
+
+    @Override
+    public void startSpan(String spanName, Level level) {
+        Span span = createSpan(spanName, getCurrentSpan(), level);
+        setCurrentSpanInContext(span);
+        addDefaultAttributes(span);
+    }
+
+    @Override
+    public void endSpan() {
+        Span currentSpan = getCurrentSpan();
+        if (currentSpan != null) {
+            currentSpan.endSpan();
+            setCurrentSpanInContext(currentSpan.getParentSpan());
+        }
+    }
+
+    @Override
+    public void addSpanAttribute(String key, String value) {
+        Span currentSpan = getCurrentSpan();
+        currentSpan.addAttribute(key, value);
+    }
+
+    @Override
+    public void addSpanAttribute(String key, long value) {
+        Span currentSpan = getCurrentSpan();
+        currentSpan.addAttribute(key, value);
+    }
+
+    @Override
+    public void addSpanAttribute(String key, double value) {
+        Span currentSpan = getCurrentSpan();
+        currentSpan.addAttribute(key, value);
+    }
+
+    @Override
+    public void addSpanAttribute(String key, boolean value) {
+        Span currentSpan = getCurrentSpan();
+        currentSpan.addAttribute(key, value);
+    }
+
+    @Override
+    public void addSpanEvent(String event) {
+        Span currentSpan = getCurrentSpan();
+        currentSpan.addEvent(event);
+    }
+
+    @Override
+    public void close() throws IOException {
+        ((Closeable) tracingTelemetry).close();
+    }
+
+    // Visible for testing
+    Span getCurrentSpan() {
+        return tracerContextStorage.get(CURRENT_SPAN);
+    }
+
+    private Span createSpan(String spanName, Span parentSpan, Level level) {
+        return spanFactory.createSpan(spanName, parentSpan, level);
+    }
+
+    private void setCurrentSpanInContext(Span span) {
+        tracerContextStorage.put(CURRENT_SPAN, span);
+    }
+
+    private void addDefaultAttributes(Span span) {
+        span.addAttribute(SPAN_ID, span.getSpanId());
```

Comment:
Why we need to duplicate SPAN_ID / TRACE_ID / SPAN_NAME (+ parent) as attributes ?  Those are part of span's metadata, right?

Reply:
We had added this for debugging, but agree we dont need it. Remove them in new revision